tensorflow/python/ops/ragged/ragged_embedding_ops.py

```diff
@@ -0,0 +1,242 @@
+def safe_embedding_lookup_sparse(embedding_weights,
+                                 sparse_ids: ragged_tensor.Ragged,
+                                 sparse_weights=None,
+                                 combiner="mean",
+                                 default_id=None,
+                                 name=None,
+                                 partition_strategy="div",
+                                 max_norm=None,
+                                 allow_dense_grads=False):
+  if embedding_weights is None:
+    raise ValueError(f"Missing embedding_weights {embedding_weights}.")
+  if isinstance(embedding_weights, variables.PartitionedVariable):
+    embedding_weights = list(embedding_weights)  # get underlying Variables.
+  if not isinstance(embedding_weights, list):
+    embedding_weights = [embedding_weights]
+  if len(embedding_weights) < 1:
+    raise ValueError(f"Missing embedding_weights {embedding_weights}.")
+
+  dtype = sparse_weights.dtype if sparse_weights is not None else None
+  embedding_weights = [
+      w if (isinstance(w, resource_variable_ops.ResourceVariable)
+            and dtype in (None, w.dtype))
+      else ops.convert_to_tensor(w, dtype=dtype)
+      for w in embedding_weights
+  ]
+
+  with ops.name_scope(name, "embedding_lookup", embedding_weights +
+                      [sparse_ids, sparse_weights]) as scope:
+
+    # Prune invalid ids and weights.
+    sparse_ids, sparse_weights = _prune_invalid_ids_ragged(sparse_ids,
+                                                            sparse_weights)
+    if combiner != "sum":
+      sparse_ids, sparse_weights = _prune_invalid_weights_ragged(
+          sparse_ids, sparse_weights)
+    sparse_ids, is_row_empty = ragged_array_ops.fill_empty_rows(
+      sparse_ids, default_id or 0)
+    if sparse_weights is not None:
+      sparse_weights, _ = ragged_array_ops.fill_empty_rows(
+          sparse_weights, 1.0)
+
+    result = embedding_lookup_sparse(
+        embedding_weights,
+        sparse_ids,
+        sparse_weights,
+        combiner=combiner,
+        partition_strategy=partition_strategy,
+        name=None if default_id is None else scope,
+        max_norm=max_norm,
+        allow_dense_grads=allow_dense_grads)
+
+    if default_id is None:
+      # Broadcast is_row_empty to the same shape as embedding_lookup_result,
+      # for use in Select.
+      is_row_empty = array_ops.tile(
+          array_ops.reshape(is_row_empty, [-1, 1]),
+          array_ops.stack([1, array_ops.shape(result)[1]]))
+
+      result = array_ops.where(
+          is_row_empty, array_ops.zeros_like(result), result, name=scope)
+
+    return result
+
+
+def _prune_invalid_ids_ragged(ids, weights):
+  """Prune invalid IDs (< 0) from the input ids and weights."""
+  is_id_valid = math_ops.greater_equal(ids.values, 0)
+  if weights is not None:
+    is_id_valid = math_ops.logical_and(
+        is_id_valid,
+        array_ops.ones_like(weights.values, dtype=dtypes.bool))
```

Comment:
Why do we need this line? I thought ids and weights have the same shape. If so, the is_id_valid should stay the same after the logical_and?

Assuming we don't need this line, the function could be refactor something like this
```
mask = ids >= 0
ids = tf.ragged.boolean_mask(ids, mask)
weights = tf.ragged.boolean_mask(weights, mask)
return ids, weights
```